In [1]:
from collections import OrderedDict
from diviner import file_utils as fu
import datetime as dt
import pandas as pd
import os

In [2]:
timestr = '2013040903'

In [3]:
rdr = fu.RDRReader(os.path.join(fu.rdrdatapath, timestr+'_RDR.TAB.zip'))

In [167]:
df = rdr.read_df(do_parse_times=False)

In [168]:
df

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5315436 entries, 0 to 5315435
Columns: 33 entries, date to qmi
dtypes: float64(24), int64(7), object(2)

In [170]:
from diviner.write_rdr20_file import *

In [171]:
set_orientation(df)

In [172]:
df.o.value_counts()

0    4846716
2     341712
3     127008
dtype: int64

In [173]:
set_view(df)

In [174]:
df.v.value_counts()

0    4846716
4     293328
6     127008
5      48384
dtype: int64

In [176]:
set_instrument(df)

In [177]:
df.i.value_counts()

0    5315436
dtype: int64

In [178]:
set_moving(df)

In [179]:
df.m.value_counts()

0    5094684
1     220752
dtype: int64

In [180]:
set_q(df)

In [181]:
df.q.value_counts()

0    5315436
dtype: int64

In [182]:
set_p(df)

In [183]:
df.p.value_counts()

2    5315436
dtype: int64

In [184]:
set_e(df)

In [185]:
df.e.value_counts()

2    5315436
dtype: int64

In [189]:
set_n(df)

In [190]:
df.n.value_counts()

0    5315436
dtype: int64

In [191]:
import diviner.write_rdr20_file as writer

In [193]:
reload(writer)

<module 'diviner.write_rdr20_file' from '/u/paige/maye/src/diviner/write_rdr20_file.py'>

In [194]:
writer.process_qmi(df)

In [196]:
for flag in 'zthdsab':
    print flag
    print df[flag].value_counts()

z
0    5315436
dtype: int64
t
0    5315436
dtype: int64
h
0    5315436
dtype: int64
d
0    5315436
dtype: int64
s
0    5314680
1        756
dtype: int64
a
0    5315436
dtype: int64
b
0    5315436
dtype: int64


Orientation
===========

In [7]:
arr = abs(df_ch.af.unique())
arr

array([110, 350, 350, 480, 480, 110, 360, 360])

In [9]:
truncated.astype('int')

array([1, 3, 3, 4, 4, 1, 3, 3])

In [16]:
trunc(df.af / 100).unique()

array([ 1., -4.,  3., -5.,  4., -2.])

In [27]:
trunc?

In [30]:
df['o'] = np.trunc(np.abs(df.af) / 100.0)

In [31]:
df.o.unique()

array([ 1.,  3.,  4.])

In [38]:
(abs(df.af) - df.o *100).unique()

array([ 10.,  50.,  80.,  60.])

In [53]:
randint(0,9,size=len(arr))

array([4, 4, 7, 5, 1, 5, 3, 4])

In [54]:
arr = arr + randint(0,9,size=len(arr))

In [55]:
arr

array([117, 354, 351, 483, 483, 117, 367, 361])

In [62]:
arr // 100

array([1, 3, 3, 4, 4, 1, 3, 3])

View
====

In [8]:
mapper = {0: 8,
          1: 0,
          2: 1,
          5: 4,
          6: 5,
          8: 6,
          9: 9}


In [29]:
tmp = ((np.abs(df.af) % 100) / 10)

In [33]:
df['v'] = 0

In [34]:
for oldval,newval in mapper.iteritems():
    df['v'][tmp ==oldval] = newval

In [35]:
df.v.unique()

array([0, 4, 6, 5])

In [36]:
tmp.unique()

array([1, 5, 8, 6])

Instrument
==========

In [37]:
df['i'] = np.abs(df.af) % 10

In [38]:
df['i'].unique()

array([0])

Moving
======

In [39]:
df['m'] = 0
df['m'][df.af < 0] = 1

In [42]:
df.m.value_counts()

0    5094684
1     220752
dtype: int64

In [41]:
df.af.value_counts()

 110    4792284
 350     196560
-350      96768
 480      90720
-110      54432
-480      36288
-360      33264
 360      15120
dtype: int64

In [43]:
df.af[df.af < 0].size

220752

qca
===

In [2]:
from diviner.flags import CalibFlag, GeomFlag, MiscFlag

In [48]:
cflag = CalibFlag(0)

In [58]:
for i in [0,1,2,4,8]:
    cflag.value = i
    cflag.set_members()
    print cflag

interp_marker_oob  :  False
nearest_marker     :  False
nearest_marker_oob :  False
constants_only     :  False

interp_marker_oob  :   True
nearest_marker     :  False
nearest_marker_oob :  False
constants_only     :  False

interp_marker_oob  :  False
nearest_marker     :   True
nearest_marker_oob :  False
constants_only     :  False

interp_marker_oob  :  False
nearest_marker     :  False
nearest_marker_oob :   True
constants_only     :  False

interp_marker_oob  :  False
nearest_marker     :  False
nearest_marker_oob :  False
constants_only     :   True



In [55]:
print cflag

interp_marker_oob  :   True
nearest_marker     :  False
nearest_marker_oob :  False
constants_only     :  False



[0,1,2,4,8,9] is almost [0,1,2,3,8,9], so i might as well just copy the old values

qge
===

In [21]:
geomflag = GeomFlag(24)
print geomflag

pointing_res  :  False
ephem_res     :  False
pointing_def  :  False
ephem_def     :   True
pointing_pred :   True
ephem_pred    :  False
pointing_none :  False
ephem_none    :  False



In [7]:
geomflag.bits_data

(('pointing_res', 1),
 ('ephem_res', 2),
 ('pointing_def', 4),
 ('ephem_def', 8),
 ('pointing_pred', 16),
 ('ephem_pred', 32),
 ('pointing_none', 64),
 ('ephem_none', 128))

In [15]:
geomflag.set_bit(32)

36


In [88]:
geomflag.ephem_def

False

In [97]:
values =array([0,1,4,16,64])
sqrt(values)

array([ 0.,  1.,  2.,  4.,  8.])

In [89]:
def process_qge(val):
    geomflag = GeomFlag(val)
    if geomflag.pointing_def:
        return 2
    elif geomflag.pointing_res:
        return 3
    elif geomflag.pointing_pred:
        return 1
    else:
        return 0

In [122]:
geomflag.dic

OrderedDict([('pointing_res', 1), ('ephem_res', 2), ('pointing_def', 4), ('ephem_def', 8), ('pointing_pred', 16), ('ephem_pred', 32), ('pointing_none', 64), ('ephem_none', 128)])

In [143]:
bitwise_and(df.qge.head(),32)

0    0
1    0
2    0
3    0
4    0
Name: qge, dtype: int64

In [129]:
df['p']=0

In [136]:
def set_p(df):
    df['p'] = 0
    df.p[bitwise_and(df.qge, 16) !=0] = 1
    df.p[bitwise_and(df.qge, 4) !=0] = 2
    df.p[bitwise_and(df.qge, 1) !=0] = 3

In [137]:
set_p(df)

In [135]:
df.p.unique()

array([2])

In [92]:
%timeit df.qge.map(process_qge)

1 loops, best of 3: 1min 14s per loop


qmi
===

In [156]:
df.qmi.value_counts()

0      5094537
128     220143
192        609
64         147
dtype: int64

In [155]:
miscflag.dic

OrderedDict([('rfu0', 1), ('eclipse', 2), ('turn_on_transient', 4), ('abnormal_instrument_state', 8), ('abnormal_instrument_temp_drift', 16), ('noise', 32), ('ch1_saturation', 64), ('moving', 128)])

In [157]:
miscflag = MiscFlag(192)

In [158]:
print miscflag

rfu0                           :  False
eclipse                        :  False
turn_on_transient              :  False
abnormal_instrument_state      :  False
abnormal_instrument_temp_drift :  False
noise                          :  False
ch1_saturation                 :   True
moving                         :   True



In [161]:
df['s'] = 0
df.s[bitwise_and(df.qmi, 64) !=0] = 1

In [162]:
df.s.value_counts()

0    5314680
1        756
dtype: int64

In [166]:
bitwise_and(df.qmi, 64).value_counts()

0     5314680
64        756
dtype: int64

In [64]:
reload(fu)

<module 'diviner.file_utils' from '/u/paige/maye/src/diviner/file_utils.pyc'>

In [8]:
pump = fu.L1ADataPump(timestr)

In [9]:
newdf = fu.read_l1a_data(pump.fnames[0])

In [10]:
newdf = fu.prepare_data(newdf)

In [11]:
fu.define_sdtype(newdf)

In [25]:
newdf[newdf.is_moving].count()

q             1168
date          1168
utc           1168
sclk          1168
sounding      1168
from_pkt      1168
pkt_count       73
safing          73
safed           73
freezing        73
frozen          73
rolling         73
dumping         73
moving        1168
temp_fault      73
...
b3_18                 1168
b3_19                 1168
b3_20                 1168
b3_21                 1168
sdtype                1168
calib_block_labels    1168
sv_block_labels       1168
bb_block_labels       1168
st_block_labels       1168
is_spaceview          1168
is_bbview             1168
is_stview             1168
is_moving             1168
is_stowed             1168
is_calib              1168
Length: 259, dtype: int64

In [26]:
newdf.from_pkt.groupby(newdf.from_pkt.isnull()).size()

from_pkt
False       28128
dtype: int64

confirmed that the 'moving' definition of my calibration and what is in 'af' is equivalent: Same count of 1168 values:

In [19]:
df_ch[(df_ch.af < 0) & (df_ch.det==1)].count()

date         1168
utc          1168
jdate        1168
orbit        1168
sundist      1168
sunlat       1168
sunlon       1168
sclk         1168
sclat        1168
sclon        1168
scrad        1168
scalt        1168
el_cmd       1168
az_cmd       1168
af           1168
orientlat    1168
orientlon    1168
c            1168
det          1168
vlookx       1168
vlooky       1168
vlookz       1168
radiance     1168
tb           1168
clat         1168
clon         1168
cemis        1168
csunzen      1168
csunazi      1168
cloctime     1168
qca          1168
qge          1168
qmi          1168
dtype: int64

In [8]:
df = df.drop(['qca','qge','qmi'],axis=1)